url9582d546a9_twentylemon_sudoku_generator_tgz-pJ8-sudoku_Top95J8/src/sudoku/clp/CLPBoard.java

```diff
@@ -0,0 +1,153 @@
+package sudoku.clp;
+
+import sudoku.CandidateSet;
+import java.util.ArrayList;
+import java.util.LinkedList;
+import java.util.List;
+import sudoku.SudokuBoard;
+
+/**
+ * In addition to what a SudokuBoard has, this also holds the possible values
+ * that each cell can have. SudokuBoard.getOptions(int) does this already,
+ * but this is a different approach entirely. Instead of arbitrarily assigning
+ * values to a cell, the major action is instead to eliminate a possible
+ * candidate for a cell.
+ *
+ * @author Taras Mychaskiw
+ */
+class CLPBoard {
+
+    private final List<CandidateSet> values;
+    private final SudokuBoard board;  //treat as immutable
+
+    CLPBoard(SudokuBoard board){
+        this.board = board;
+        values = new ArrayList<>(board.getNumCells());
+        for (int i = 0; i < board.getNumCells(); i++){
+            values.add(new CandidateSet(board.getSize()));
+        }
+        for (int cell = 0; cell < board.getNumCells(); cell++){
+            if (board.isSet(cell)){
+                assign(cell, board.getCell(cell));
+            }
+        }
+    }
+
+    CLPBoard(CLPBoard other){
+        this.board = other.board;
+        values = new ArrayList<>(board.getNumCells());
+        for (int i = 0; i < board.getNumCells(); i++){
+            values.add(new CandidateSet(other.getCandidateSet(i)));
+        }
+    }
+
+    CandidateSet getCandidateSet(int cell){ return values.get(cell); }
+    SudokuBoard getBoard(){ return board; }
+
+
+    /**
+     * Assigns the value to the cell. What actually happens is all other
+     * candidates for the cell are eliminated.
+     *
+     * @param cell the cell to assign a value to
+     * @param valToAssign the value to assign
+     * @return true if it was possible to assign the value
+     */
+    boolean assign(int cell, int valToAssign){
+        for (Integer value : values.get(cell)){
+            if (value != valToAssign && !eliminate(cell, value)){
+                return false;
+            }
+        }
+        return true;
+    }
+
+
+    /**
+     * Eliminates the value from the list of candidates for the cell.
+     * If any new constraints are found, they are propagated out.
+     *
+     * @param cell the cell to eliminate a candidate from
+     * @param valToRemove the candidate to remove
+     * @return true if it was possible to remove the candidate from the cell
+     */
+    boolean eliminate(int cell, int valToRemove){
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

The cyclomatic complexity of this method is 12. By comparison, 98% of the methods in the CodeGuru reference dataset have a lower cyclomatic complexity. This indicates the method has a high number of decisions and it can make the logic difficult to understand and test.

We recommend that you simplify this method or break it into multiple methods. For example, consider extracting the code blocks on lines 80-91, lines 97-114 into separate methods.